### PR TITLE
ISPN-9688 Internal REST server modularization

### PR DESCRIPTION
**Pull request by gustavonalle**
https://issues.jboss.org/browse/ISPN-9688

* This should keep the REST server behavior intact;
* Perf wise, it should be roughly the same. I measured a ~5% increase in throughput and ~5-10% decrease in object allocation, but those are likely noise, let's see what RadarGun says.

### CODE REVIEW

**Comment by gustavonalle** [comment]
run performance tests please
cs-rest

**Comment by gustavonalle** [comment]
@ryanemerson I can sure add more javadocs, any particular one that needs more clarity?

Probably looking at the framework tests themselves or better, the [CacheResource](https://github.com/infinispan/infinispan/blob/0d344311b572cde316a7c42a7b03c7ff88e03366/server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java#L54), it's a good start to review.

Regarding versioning, the idea is to use https://issues.jboss.org/browse/ISPN-8535?focusedCommentId=13536259&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-13536259

File: server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+      if (returnBody) responseBuilder.entity(value);
+   }
+
+   private <V> String calcETAG(V value) {
+      return String.valueOf(hashFunc.hash(value));
+   }
+
+   private RestResponse putInCache(NettyRestResponse.Builder responseBuilder, boolean useAsync, AdvancedCache<
+         Object, Object> cache, Object key,
+                                   byte[] data, Long ttl, Long idleTime, Optional<Object> prevCond) {
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java]: `prevCond` param shouldn't be optional, just handle if null.

File: server/rest/src/main/java/org/infinispan/rest/framework/impl/SimpleRequest.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+import io.netty.handler.codec.http.QueryStringDecoder;
+
+/**
+ * @since 10.0
+ */
+public class SimpleRequest implements RestRequest {
+   private final Method method;
+   private final String path;
+   private Map<String, String> headers;
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/framework/impl/SimpleRequest.java]: `headers` is never used

File: server/rest/src/main/java/org/infinispan/rest/NettyRestResponse.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+      @Override
+      public Builder eTag(String tag) {
+         response.headers().set(HttpHeaderNames.ETAG, tag);
+         return this;
+      }
+
+      @Override
+      public int getStatus() {
+         return response.getStatus().code();
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/NettyRestResponse.java]: `response.status().code()`

File: server/rest/src/main/java/org/infinispan/rest/NettyRestResponse.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+         return this;
+      }
+
+      public Builder clusterServerAddress(String serverAddress) {
+         response.headers().set(CLUSTER_SERVER_ADDRESS_HEADER, serverAddress);
+         return this;
+      }
+
+      public HttpVersion getHttpVersion() {
+         return response.getProtocolVersion();
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/NettyRestResponse.java]: `response.protocolVersion()`

File: server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceQueryAction.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      return JsonQueryReader.getQueryFromJSON(byteContent);
+   }
+
+   private String getParameterValue(RestRequest restRequest, String name) {
+      List<String> values = restRequest.parameters().get(name);
+      if (values == null) return null;
+      return values.iterator().next();
+   }
+
+   private NettyRestResponse error(RestRequest restRequest, HttpResponseStatus status, String message, String cause) {
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceQueryAction.java]: `restRequest` param is ignored and `status` is always`BAD_REQUEST`

File: server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceQueryAction.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      ContentSource contents = restRequest.contents();
+      byte[] byteContent = contents.rawContent();
+      if (byteContent == null || byteContent.length == 0) throw new IOException();
+      return JsonQueryReader.getQueryFromJSON(byteContent);
+   }
+
+   private String getParameterValue(RestRequest restRequest, String name) {
+      List<String> values = restRequest.parameters().get(name);
+      if (values == null) return null;
+      return values.iterator().next();
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/resources/CacheResourceQueryAction.java]: return `values == null ? null : values.iterator().next();`

File: server/rest/src/main/java/org/infinispan/rest/NettyRestRequest.java [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+   public boolean getPerformAsyncHeader() {
+      return getHeaderAsBoolean(PERFORM_ASYNC_HEADER);
+   }
+
+   @Override
+   public Long getTimeToLiveSecondsHeader() {
+      return getHeaderAsLong(TTL_SECONDS_HEADER);
+   }
+
+   public FullHttpRequest getFullHttpRequest() {
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/NettyRestRequest.java]: Never used

File: server/rest/src/main/java/org/infinispan/rest/NettyRestRequest.java [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+import io.netty.handler.codec.http.QueryStringDecoder;
+
+/**
+ * A {@link RestRequest} backed by Netty.
+ *
+ * @since 10.0
+ */
+public class NettyRestRequest implements RestRequest {
+
+   protected final static Log logger = LogFactory.getLog(NettyRestRequest.class, Log.class);
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/NettyRestRequest.java]: private?

File: server/rest/src/main/java/org/infinispan/rest/framework/impl/PathItem.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
+package org.infinispan.rest.framework.impl;
+
+/**
+ * Represents each hop of a path. Example: /rest/a/b is formed by 3 path items: rest, a, b.
+ *
+ * @since 10.0
+ */
+public abstract class PathItem {
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/framework/impl/PathItem.java]: package-private?

File: server/rest/src/main/java/org/infinispan/rest/framework/impl/RestRequestBuilder.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.Map;
+
+import org.infinispan.rest.framework.ContentSource;
+import org.infinispan.rest.framework.Method;
+import org.infinispan.rest.framework.RestRequest;
+
+/**
+ * @since 10.0
+ */
+public interface RestRequestBuilder<B extends RestRequestBuilder<B>> {
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/framework/impl/RestRequestBuilder.java]: package-private?

File: server/rest/src/main/java/org/infinispan/rest/framework/impl/SimpleRestResponse.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+   }
+
+   @Override
+   public Object getEntity() {
+      return builder.getEntity();
+   }
+
+   public static class Builder implements RestResponseBuilder<Builder> {
+
+      protected Map<String, Object> headers = new HashMap<>();
```
Comment (ryanemerson) [inline server/rest/src/main/java/org/infinispan/rest/framework/impl/SimpleRestResponse.java]: private?

**Comment by ryanemerson** [comment]
@gustavonalle Looking at the framework interfaces a bit more and it's pretty self explanatory with the mapping to http conventions, so I take back my javadocs comment.

File: server/rest/src/main/java/org/infinispan/rest/framework/ResourceHandler.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+ * Handles all the logic related to a REST resource.
+ *
+ * @since 10.0
+ */
+public interface ResourceHandler {
+
+   /**
+    * Return the {@link Invocations} handled by this ResourceHandler.
+    */
+   Invocations getInvocations();
```
[the remark this answers is not in the document]
Comment (gustavonalle) [reply server/rest/src/main/java/org/infinispan/rest/framework/ResourceHandler.java]: Could be, but the builder that creates ```Invocations``` abstracts the collection construction internally, the end result is a more compact code

File: server/rest/src/main/java/org/infinispan/rest/framework/impl/ResourceManagerImpl.java [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+   private final ResourceNode resourceTree;
+   private final String rootPath;
+
+   public ResourceManagerImpl(String rootPath) {
+      this.rootPath = rootPath;
+      this.resourceTree = new ResourceNode(new StringPathItem(rootPath), null);
+   }
+
+   ResourceManagerImpl() {
```
[the remark this answers is not in the document]
Comment (gustavonalle) [reply server/rest/src/main/java/org/infinispan/rest/framework/impl/ResourceManagerImpl.java]: np

File: server/rest/src/main/java/org/infinispan/rest/framework/ResourceHandler.java [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+   String getName();
```
[the remark this answers is not in the document]
Comment (gustavonalle) [reply server/rest/src/main/java/org/infinispan/rest/framework/ResourceHandler.java]: I was going to use it to generate documentation, but until then it's better to remove it.

File: server/rest/src/main/java/org/infinispan/rest/framework/impl/ResourceManagerImpl.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+      this("root");
+   }
+
+   @Override
+   public void registerResource(ResourceHandler handler) {
+      handler.getInvocations().forEach(invocation -> {
+         Set<String> paths = invocation.paths();
+         paths.forEach(path -> {
+            if (path.startsWith("/")) path = path.substring(1);
```
[the remark this answers is not in the document]
Comment (gustavonalle) [reply server/rest/src/main/java/org/infinispan/rest/framework/impl/ResourceManagerImpl.java]: ok

File: server/rest/src/main/java/org/infinispan/rest/framework/impl/ResourceNode.java [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+               VariablePathItem vpi = (VariablePathItem) pi.getKey();
+               Map<String, String> vars = PathInterpreter.resolveVariables(vpi.getExpression(), path);
+               if (!vars.isEmpty()) {
+                  variables.putAll(vars);
+                  return true;
+               }
+               return false;
+            }).findFirst();
+
+      return first.map(Map.Entry::getValue).orElse(null);
```
[the remark this answers is not in the document]
Comment (gustavonalle) [reply server/rest/src/main/java/org/infinispan/rest/framework/impl/ResourceNode.java]: np

File: server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+   private Throwable getRootCauseException(Throwable re) {
+      if (re == null) return null;
+      Throwable cause = re.getCause();
+      if (cause != null)
+         return getRootCauseException(cause);
+      else
+         return re;
+   }
+
+   MediaType tryNarrowMediaType(MediaType negotiated, AdvancedCache<?, ?> cache) {
```
[the remark this answers is not in the document]
Comment (gustavonalle) [reply server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java]: WRT privates...Not sure, maybe this class will be extended to provide the CacheResource v2 :)

File: server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+      MediaType storageMediaType = cache.getValueDataConversion().getStorageMediaType();
+
+      if (storageMediaType == null) return negotiated;
+      if (storageMediaType.equals(MediaType.APPLICATION_OBJECT)) return TEXT_PLAIN;
+      if (storageMediaType.match(MediaType.APPLICATION_PROTOSTREAM)) return APPLICATION_JSON;
+
+      return negotiated;
+   }
+
+   MediaType negotiateMediaType(String accept, String cacheName) throws UnacceptableDataFormatException {
```
[the remark this answers is not in the document]
Comment (gustavonalle) [reply server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java]: WRT privates...Not sure, maybe this class will be extended to provide the CacheResource v2 :)

File: server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java (hunk 1 above)
Comment (gustavonalle) [reply server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java]: Good catch, I forgot to remove this Optional

File: server/rest/src/main/java/org/infinispan/rest/framework/impl/SimpleRequest.java (hunk 2 above)
Comment (gustavonalle) [reply server/rest/src/main/java/org/infinispan/rest/framework/impl/SimpleRequest.java]: ```SimpleRequest``` is the default implementation of ```RestRequest``` not tied to any particular underlying http library, this allows to test the base frameworks classes in isolation. As it evolves, I expect it to the used in future tests

File: server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (gustavonalle) [reply server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java]: Changed my mind, private it is

**Comment by gustavonalle** [comment]
> @gustavonalle Looking at the framework interfaces a bit more and it's pretty self explanatory with the mapping to http conventions, so I take back my javadocs comment.

Good, that was the idea, a minimalist, dependency-free yet modular way of restfying everything :)

**Comment by gustavonalle** [comment]
@ryanemerson Addressed reviews

**Pull request merged**